098-using-image-volumes-to-improve-extensibility-of-Strimzi-operands.md

```diff
@@ -0,0 +1,335 @@
+kind: KafkaConnect
+metadata:
+  # ...
+spec:
+  # ...
+  plugins:
+    - name: echo-sink-connector
+      artifacts:
+        - type: image
+          reference: ghcr.io/scholzj/echo-sink@sha256:7078b3ccbc0d6e76fefb832dd6e4bb6b704f83611428c9d5b060ab0b6c3e8712
+          pullPolicy: Always
+```
+
+The software in this case will be defined as the following set of container images:
+* `quay.io/scholzj/kafka@sha256:fd0a060b512240c6dc45e90de26fd738fbbcd470dbb1a3c4c8ec808392db6159` as the _Strimzi_ container image from my prototype build
+* `ghcr.io/scholzj/echo-sink@sha256:7078b3ccbc0d6e76fefb832dd6e4bb6b704f83611428c9d5b060ab0b6c3e8712` as the Kafka Connect connector image
+
+Each of these are immutable container images and clearly define the software being run in a given operand.
+
+### Error handling
+
+Strimzi cannot easily detect if the `ImageVolume` feature gate is enabled and if image volumes are supported in a given Kubernetes cluster.
+While we will use the documentation to warn users that this feature can be used only in Kubernetes cluster where it is supported, it might happen that users try to use it even when not supported by Kubernetes.
+In that case, Kubernetes will ignore the image volume and its volume mount and create the Pod without it (tested with Minikube on Kubernetes 1.31 and 1.32, and with OpenShift 4.18).
+As a result, the plugin will be missing and the user has to choose another way how to add it.
+
+In case the image volumes are enabled in Kubernetes but not supported by the container runtime, the container will end up in the `CreateContainerError` state with the following message:
+
+```yaml
+  status:
+    state:
+      waiting:
+        message: 'Error response from daemon: invalid volume specification: '':/mnt/echo-sink-plugin/:ro'''
+        reason: CreateContainerError
+```
+
+In such case, the user will need to edit the related Strimzi custom resource, remove the parts using the image volumes and apply the changes.
+Afterwards, the Strimzi operator will recover and recreate the Pod with a valid configuration.
+
+Use of a feature gate to prevent the use of image volumes on unsupported cluster was considered.
+But since it is expected that we will be seeing Kubernetes clusters which do not support image volumes for a very long time, the feature gate which would be for a long time in alpha phase and would need to be always enabled does not seem to make sense.
+
+### Testing strategy
+
+The functionality from this proposal will be initially tested only using unit tests.
+This should be sufficient in terms of test coverage given we only create the right Pod definition and the rest is up to Kubernetes.
+Introducing a system test right now would mean that such a test would be executable only on very specific Kubernetes versions with specific configurations and would not add much value.
+System tests using image volumes might be added at a later stage once the image volume support reaches more Kubernetes clusters.
```

Comment:
This has to be tracked with an issue imho. We should create one as part of this proposal being approved.

Reply:
I'm completely fine to open an issue for that when I open the PR for this proposal. However, as I said:
* I do not think we need STs for this as it is in general Kubernetes feature. We do not have STs for many other features that are Kubernertes responsibility.
* I think it would be pretty hard to track when the issue is _ready to be worked on_. It will be elft there for a long time and forgotten even after we add the tests.

So I do not think that is needed. But if you want it, no problem, I can open it.